content/en/blog/_posts/2023-12-13-Canary-Deployment-With-Kubernetes/index.md

```diff
@@ -0,0 +1,282 @@
+as a way to define the rules for how request to a particular service should be handled
+within the Istio service mesh.
+
+You can use the virtual service to route a percentage of traffic to the new version of an
+app for testing purposes, apply different routing rules based on request headers,
+or define fault injection policies.
+
+Add the following lines to your istio.yaml file.
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: canary-app-virtual-service
+spec:
+  hosts:
+    - "*"
+  gateways:
+    - canary-gateway
+  http:
+  - route:
+    - destination:
+        host: canary-app-service.default.svc.cluster.local
+        subset: v1
+      weight: 90
+    - destination:
+        host: canary-app-service.default.svc.cluster.local
+        subset: v2
+      weight: 10
+```
+
+You can see that here the `canary-gateway` is specified in the virtual service. Specifying
+the gateway makes the service accessible from outside the cluster. Here is a breakdown of
+the `route.destination.host` field.
+
+- `canary-app-service`: This is the name of the service of the app.
+- `default`: This is the namespace in which the service resides. In this case, it's the default namespace.
+- `svc`: It stands for "service" and is a standard suffix for Kubernetes services.
+- `cluster.local`: This is the default cluster domain in Kubernetes.
+
+The `weight` field defines the routing weight of the requests. The `subset` field allows
+you to specify a destination for the requests specified via a `DestinationRule` resource.
+Create and add a `DestinationRule` resource into the istio.yaml file.
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: DestinationRule
+metadata:
+  name: canary-app
+spec:
+  host: canary-app-service
+  subsets:
+  - name: v1
+    labels:
+      version: v1
+  - name: v2
+    labels:
+      version: v2
+```
+
+The `DestinationRule` resource defines `subsets`. The `subsets` allow you to
+define a group of pods as a destination for the incoming requests using pod labels.
+
+Now apply the istio.yaml using the following command.
+
+```shell
+kubectl apply -f istio.yaml
+```
+
+After applying the istio.yaml use the following command to get the external IP of the
+Istio ingress gateway load balancer.
+
+```shell  
+kubectl get svc -n istio-system
+```
+
+Note: If you are using the minikube, and the external IP of the load balancer is in the
+pending state use the `minikube tunnel` command. Running this command will provide an
+external IP for your load balancer. And if you are using the `kind` to create the
+kubernetes cluster. You would need to install the [MetalLB](https://metallb.universe.tf/installation/).
+
+Copy the external IP of the load balancer and paste that into your browser, and specify
+the port `80`. You will see the output of the app. And if you refresh the page several
+times you'd be able to see that the traffic is distributed in a 9:1 ratio. You can
```

Comment:
```suggestion
times you'll see that the traffic is distributed in a 9:1 ratio. You can
```

Reply:
Thanks for the suggestions, I will get back to it in my free time :sweat_smile: